### DISCUSSION TITLE
Tearing my hair out

### DISCUSSION THREAD

**Discussion by marine1983boots** [kyleskom/NBA-Machine-Learning-Sports-Betting#132] [General]
Hi,

I'm very new to python and really need help getting this to work on my PC.

I'm trying to figure out what I'm doing wrong. I can get it to work in google colab. On my machine, its nearly went out the window a few times. 

I've copied the zip file extracted to my drive.

I installed Pycharm and python.

This is where my problems start.

The requirements txt file has versions that Pycharm cant install. the versions of tensor flow 2.6 cant be found.

![image](https://user-images.githubusercontent.com/127334214/223783734-69a7dc76-0989-47ae-a95b-26d4f4331755.png)
 

When I run the command

python .\main.py -A -odds==fanduel

I get a string of errors on the screen. I need to understand why it works on colab and not my computer. Any help is appreciated.

![image](https://user-images.githubusercontent.com/127334214/223780393-baa0df59-1b44-4f31-b96b-ba1c1585bfa5.png)

**Comment by nealmick** [comment]
I think using a virtual environment might fix this:
https://packaging.python.org/en/latest/guides/installing-using-pip-and-virtual-environments/
Then install requirements while environment is active.

**Reply by marine1983boots** [reply]
Hi,

Pycharm has setup a virtual directory for me. Thanks

**Comment by kyleskom** [comment]
What version on Python are you using?

**Reply by marine1983boots** [reply]
The python version im running is 3.10, which I believe is compatible with TensorFlow 2.6.0.

For some reason the pip install will not find 2.6. It goes down to 2.8 only. Do I need to downgrade pip also?

**Comment by marine1983boots** [comment]
Ok, got a little further and worked out that the odds are not being pulled back. It seems to error on the Sbroddsprovider.py

![image](https://user-images.githubusercontent.com/127334214/223966704-b36edc1b-056e-4f3e-808e-6d1e339fd673.png)

Then this error in the console.

![image](https://user-images.githubusercontent.com/127334214/223966834-ca38c2f9-9c6e-452f-a62b-b8350594c4f0.png)

**Comment by marine1983boots** [comment]
Ok so I started again. I've created a new pycharm instance and copied the install instructions verbatim. 

Its definitely something to do with the odds as they are empty

![image](https://user-images.githubusercontent.com/127334214/224019721-00908e22-3f5b-4eed-bb57-03d24c09626e.png)

**Reply by nealmick** [reply]
Seems to be working fine for me.
![Screen Shot 2023-03-09 at 8 09 46 AM](https://user-images.githubusercontent.com/33704976/224032957-59cf86e8-89d5-4501-80d2-2df671b53b6e.png)

In my opinion PyCharm is quite a heavy IDE.  Lots of features and configuration settings, especially around project setup.... Just for simplicity and getting things to work, I would try installing using a regular terminal (or your windows emulated equivalent), get stuff working there.... then maybe open in PyCharm and use more as a text editor approach.....

**Reply by marine1983boots** [reply]
I'm new to python in general. Reason why I choose Pycharm was because it acts like windows. I can change directory like windows and run from the root etc.

Python is so different for me. 

I've just launched IDLE and ran the git clone and I get an error. Flying computers inbound. lol I know its me, but Python is hard work if coming from a windows background.

![image](https://user-images.githubusercontent.com/127334214/224058552-0e5a2636-79f1-49c3-ad89-c6ceece08b34.png)

**Reply by nealmick** [reply]
yea windows can get a bit aggravating, i'm a big mac/linux guy myself.  In the screen shot you posted, its the IDLE python interpreter.  Its easy to confuse this with something like terminal or command prompt.  But its completely different.  Its basically a terminal but only for python... its the same thing you get when you type python3 into terminal.... the live python interpreter....  What i was referring to somthing like terminal for windows. 
https://apps.microsoft.com/store/detail/windows-terminal/9N0DX20HK701?hl=en-us&gl=us&rtc=1
This should enable a more native like linux terminal experience.
Also if you still cant get running I have a similar project that is hosted here nbadata.cloud

**Comment by hendrixguar** [comment]
@marine1983boots I recently propped this up in Windows with only basic Python knowledge, so maybe I can help. As kyleskom alluded to, I wouldn't try running this on a version of Python later than 3.8. I noticed in your most recent screenshot you are using 3.10. The rub is that you can't download 3.8 as an executable anymore. But good news! There is a workaround and I actually found it quite easy.

1. For straightforward results, uninstall any versions of Python if at all possible. Use the command line (--version) to ensure you are working with a clean slate.
2. Follow the instructions here to install pyenv-win: https://github.com/pyenv-win/pyenv-win This is a fork of pyenv (for Unix), and it's a single-use utility which allows you to install and run multiple versions of Python simultaneously, directly from the command line. You can also use this to get a later version for other projects.
3. Use pyenv-win to install Python 3.8.10, and also use it to ensure the version you're using is 3.8. If you get a different result, you have a conflict with a previous installation.
4. Follow the exact command line instructions in the Usage section for this project. When I installed in this manner I no longer had the issue with tensorflow version, for example.

**Reply by marine1983boots** [reply]
So grateful for the reply. I don't like to be beaten by computers ha

I've found the issue with the help of ChatGPT.

When the sportsbook is called, there is no "fanduel" in the returned odds. Hence it was throwing the error. The only results I get back are below. I'm from the UK and wonder if the website it scrapes from is geo-locating me to UK bookies. 

![image](https://user-images.githubusercontent.com/127334214/224255658-7345ebc3-a6be-44af-aacc-39796fcf8b15.png)

This is the scraped website for me.

![image](https://user-images.githubusercontent.com/127334214/224256086-c83c1bd6-cb5a-4af7-89ce-18498b31090a.png)

Just ran it again in colab and it works. Which makes me believe so.

Just ran the code on my machine again with Bet365 as the bookie. It now works

![image](https://user-images.githubusercontent.com/127334214/224256848-ec1bc3a2-cb5e-4a6d-8272-8f381349a93f.png)


Next question how do you fix it? Do I even need to? Whats the best bookie for odds out of the above?


This was the code ChatGPT had written for me to work out what was going on.

On the init. It checks for the valid keys

`from sbrscrape import Scoreboard

class SbrOddsProvider:
    """ Abbreviations dictionary for team location which are sometimes saved with abbrev instead of full name.
    Moneyline options name require always full name
    Returns:
        string: Full location name
    """

    def __init__(self, sportsbook="fanduel"):
        sb = Scoreboard(sport="NBA")
        self.games = sb.games if hasattr(sb, 'games') else []
        self.sportsbook = sportsbook

        # Check if sportsbook is a valid key in home_ml, away_ml, and total dictionaries
        valid_keys = set()
        for game in self.games:
            valid_keys |= set(game['home_ml'].keys()) | set(game['away_ml'].keys()) | set(game['total'].keys())
        if self.sportsbook not in valid_keys:
            raise ValueError(f"Invalid sportsbook: {self.sportsbook}. Valid options are {valid_keys}")

    def get_odds(self):
        """Function returning odds from Sbr server's json content

        Returns:
            dictionary: [home_team_name + ':' + away_team_name: { home_team: money_line_odds, away_team: money_line_odds }, under_over_odds: val]
        """
        dict_res = {}
        for game in self.games:
            # Get team names
            home_team_name = game['home_team'].replace("Los Angeles Clippers", "LA Clippers")
            away_team_name = game['away_team'].replace("Los Angeles Clippers", "LA Clippers")

            money_line_home_value = money_line_away_value = totals_value = None

            # Get money line bet values
            if self.sportsbook in game['home_ml']:
                money_line_home_value = game['home_ml'][self.sportsbook]
            if self.sportsbook in game['away_ml']:
                money_line_away_value = game['away_ml'][self.sportsbook]

            # Get totals bet value
            if self.sportsbook in game['total']:
                totals_value = game['total'][self.sportsbook]

            dict_res[home_team_name + ':' + away_team_name] = {
                'under_over_odds': totals_value,
                home_team_name: {'money_line_odds': money_line_home_value},
                away_team_name: {'money_line_odds': money_line_away_value}
            }
        print(dict_res)
        return dict_res
`
